### ISSUE TITLE
Xamarin - Using SaveAsync with ViewModelBase. Issue with EditLevel mismatch.

### ISSUE THREAD

**Issue by GreatBarrier86** [MarimerLLC/cslaforum#770]
Question. I noticed that when using `ViewModelBase.SaveAsync()`, my object is first cloned, and then `ApplyEdit` is called. During `ApplyEdit`, `AcceptChanges` on my root `BusinessListBase` cascades the call to the children.

However, during the cascading `AcceptChanges` call, one of the list `BusinessBase` child object's `BusinessRules` object also gets `AcceptChanges` called on it, which causes "Edit Level mismatch in AcceptChanges." It appears to be because during cloning, `BusinessRules` does not serialize `_stackStack`, causing it to be 0 instead of 1 during `AcceptChanges`.

Line 970 of `BusinessRules` indicates that it can serialize that field `#if ANDROID || IOS`, but according to the link [here](https://github.com/MarimerLLC/cslaforum/issues/740), Lhotka indicates that "the only valid compile targets are now .NET Standard 2.0, .NET Framework 4.6.1, 4.5, and 4.0," so i assume that shouldn't be used, but if not, then how will _stateStack be serialized and usable for AcceptChanges after cloning?

Any help would be appreciated.

Thanks,
Jason

**Comment by rockfordlhotka** [comment]
> Lhotka indicates that "the only valid compile targets are now .NET Standard 2.0, .NET Framework 4.6.1, 4.5, and 4.0,"

I should have been more precise. That statement applies to `Csla.dll`, not to `Csla.Xaml.dll`, where the per-runtime compile targets remain valid.

**Comment by GreatBarrier86** [comment]
Rocky, thank you. That makes a lot more sense now. So i assume you're saying this is all we need to do, correct? 

![image](https://user-images.githubusercontent.com/29482393/60039797-9704c600-9685-11e9-8843-5d4829accecc.png)

**Comment by rockfordlhotka** [comment]
Hmm, my answer got lost 😠 

That only affects _your_ code, not the way the CSLA code is built. When you reference `CSLA-Xamarin` you are getting code that I built, and it was built with that constant set. You don't get to affect how CSLA builds - that's done before the packages are pushed to NuGet.

**Comment by GreatBarrier86** [comment]
This is where I get confused. If it was built with that constant set, then shouldn't `_stateStack` be cloned with it? That's what's causing my edit level mismatch. If i replace the clone method in my base list class with this, it works, so clearly something is getting lost during the clone process.

```
        object ICloneable.Clone()
        {
            return this;
        }
```

**Comment by rockfordlhotka** [comment]
I'll try to be clear so we can figure this out.

* `Csla.dll` is built using the `NETSTANDARD2_0` constant. There is no Xamarin-specific build of `Csla.dll`. That's where `_stateStack` is declared and used.
* `Csla.Xaml.dll` (for Xamarin.Forms) is built using the `XAMARIN` constant. There is no Android-specific or iOS-specific build of `Csla.Xaml.dll`. This is where the `ViewModel` types are declared.
* `Csla.Axml.dll` are built with the `ANDROID` constant.

You can build your code with any constants you want. Your constants will have no effect on the behavior of the CSLA assemblies from NuGet.

**Comment by GreatBarrier86** [comment]
Right. I completely understand everything you said, and yes, as far as i know, `Csla.Axml.dll` is irrelevant. What i'm saying is that it APPEARS that the "Edit state mismatch" exception i'm receiving is because the `BusinessRules._stateStack` is not being serialized during `ViewModelBase.SaveAsync`'s `Clone()` call before `ApplyEdit()` is called. Because of that, the `EditLevel` for BusinessRules is too low during `ApplyEdit()`, causing the exception. 

Also, i noticed that you're serialization of _stateStack in FieldDataManger and UndoableBase is NOT wrapped in that compiler statement, so those DO work as expected.

https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/Rules/BusinessRules.cs
![image](https://user-images.githubusercontent.com/29482393/60196047-49659600-980a-11e9-922c-eea7841337aa.png)

**Comment by rockfordlhotka** [comment]
Sounds like a potential bug for sure.

So to create a failing unit test (and I'd like your confirmation of these steps) I'd:

1. Create an editable root instance
1. Elevate that object's editlevel via `BeginEdit`
1. Clone the object
1. Call `ApplyEdit` on the clone
1. Save the clone

Somewhere in there it should fail?

**Comment by GreatBarrier86** [comment]
In my scenario, a root item is the list itself, so in order for it to fail, it needs a child item added to the list, which i demonstrated below. This example fails for me. `CreateChild` simply creates a new `Diary`, marks it as a child, and adds it to the list.

```
            DiaryList list = new DiaryList();
            list.BeginEdit();
            list.CreateChild();
            list = (DiaryList)((ICloneable)list).Clone();
            list.ApplyEdit();
```
